datasets/utils/secret.py

```diff
@@ -0,0 +1,213 @@
+import base64
+import json
+import logging
+import os
+from collections import defaultdict
+from dataclasses import dataclass
+from typing import Dict, Union
+
+import boto3
+from kubernetes import client, config
+from tenacity import (
+    Retrying,
+    retry_if_exception,
+    stop_after_attempt,
+    wait_fixed,
+)
+
+
+logger = logging.getLogger(__name__)
+
+# Ensure we return a non-None value
+# Prefer to throw exception rather than returning None if secret can't be found
+SECRET_RETURN_TYPE = Union[str, Dict[str, str]]
+
+# Global secret cache to prevent redundant
+# secret retrieval calls to external system if possible
+secret_cache = defaultdict(dict)
+
+
+@dataclass
+class SecretFetcher:
+    """
+    Dataclass interface for secret retrieval.
+
+    Please provide one and only one of the following secret source variable:
+        kubernetes_secret_name: name of kubernetes namespaced secret,
+            in the same namespace as where this fetcher runs.
+            (extra "kubernetes" needs to be installed)
+            secret value to be pulled from k8s namespaced secret
+        aws_secret_arn: secret name or full arn of aws secret manager secret
+            use full arn for cross AWS account secret retrieval
+            secret value to be pulled from aws secret manager through boto3
+        env_var: env var name containing the secret value
+            secret value to be pulled from OS env var
+        raw_secret: raw secret value
+            secret value to be pulled directly through the variable
+
+    Optional variables:
+        key: if secret source contains multiple secrets as a dict,
+             return secret[key] if provided, otherwise return full secret
+        force_reload: ignore cached secret (if exists) and force reload from source
+
+    Interface requirement:
+        A "value" property to return the actual secret value,
+        either as a str (for single secret value) or a Dict[str, str] (for a group of secrets)
+    """
+
+    kubernetes_secret_name: str = None
+    aws_secret_arn: str = None
+    env_var: str = None
+    raw_secret: SECRET_RETURN_TYPE = None
+
+    key: str = None
+
+    force_reload: bool = False
+
+    @property
+    def value(self) -> SECRET_RETURN_TYPE:
+        """
+        Returns:
+            if key is not None: key field value of secret
+            if key is None: entire secret
+        """
+        self._variable_validation()
```

Comment:
Probably worthwhile doing this validation in the `__post_init__` method (see [these docs](https://docs.python.org/3/library/dataclasses.html#post-init-processing)). Major reason is just in case we introduce another `value` function in the future (not sure why, seems unlikely) then instead of validation on read we validate on the `constructor` which is basically the logic we want to encode.

Very minor detail and doesn't really matter for this class specifically but more generally about making the semantics of the code a little cleaner :)